### DISCUSSION TITLE
Custom netplay implementation

### DISCUSSION THREAD

**Discussion by n-at** [EmulatorJS/EmulatorJS#658] [Show and tell]
I've created a simple retro game library with built-in ability to play with EmulatorJS and built-in netplay: https://github.com/n-at/playtime 

On the user's side, netplay works like this:

1. Game host uploads a game and enables netplay in the game settings:
![1](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/3036ecd3-0989-48a0-930c-d1588cc22435)
![2](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/ee69c5aa-715c-41f4-babb-8a08ba89bbb8)

2. The game host starts the game and opens the netplay menu. There are link and QR code that can be shared with other players. No registration is required to join:
![3](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/bad2cd71-d27c-4dde-8767-e66f8489554c)
![4](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/cc8ab834-08f0-4add-929d-49e98d36d34b)

3. Other players (_clients_) start as a spectators and cannot yet control the game. The game host can assign their player id (i.e. controller number):
![5](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/7e764c64-2af9-4343-ba73-c3a0fab14adb)
![6](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/3cb59951-32e4-47e2-b6ef-d881f7de11f4)

4. Players (clients) in the netplay menu can also change their display names:
![7](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/42b884ac-516d-476d-ac59-acf672014d16)

On the technical side, netplay is implemented using WebSocket and WebRTC:

* No netplay code from EmulatorJS is used (neither client or server), everything is custom. All emulation cores should be supported.
* WebSocket used to share state (clients connecting/disconnecting, player ids and names) and exchange messages between clients (for RTC signalling). Server-side code is written on Go and is mostly straightforward: https://github.com/n-at/playtime/blob/master/web/handlers_netplay.go#L54
* Audio and video are streamed from the game host (i.e. owner of the game, the only one who has access to the game ROM and emulation settings). The video stream is captured from the canvas and the audio stream is from `AL` gain nodes. Stream is one-directional: from the game host to clients.
* Control, i.e. passing the controller state from the client to the host, including button ids and values, is done with RTC Data Channels. It is a direct connection, so input lag will be minimal. Although Data Channels are bidirectional, data is transmitted in only one direction: from the client to the game host.
* Client-side code is organized into three parts:
  * https://github.com/n-at/playtime/blob/master/assets/netplay.js - a common library where all netplay functionality is gathered (client state, WebSocket connection and RTC connection management)
  * https://github.com/n-at/playtime/blob/master/assets/netplay-ui-client.js - client code that connects the netplay library with the client UI (also contains custom virtual controller implementation)
  * https://github.com/n-at/playtime/blob/master/assets/netplay-ui-host.js - host code that connects the netplay library with game host UI (and emulator)

Maybe this can be helpful for an official netplay implementation.

**Comment by allancoding** [comment]
I really like this method of doing netplay. I have a question how is the audio captured? This is a cool project, I think this would be an awesome client to replace linuxserver.io docker image. If you would be ok with that. Another thing is how are the delays and what are the draw backs of always depending on the host. Like what is the extra performance on the host that is needed? I think the server code would need to be written in node.js but that would be not too hard. Do you have discord? If so would you mind joining the server for easier communication and collaboration. 
Thanks

**Comment by ethanaobrien** [comment]
I've tried netplay from this angle, but it's never really turned out well, er, worked.
My main concern is webrtc. The old version of netplay (emulatorjs 3.0) used webrtc, and it "worked." The main issue of using both webrtc and websockets is that the webrtc connection would seem to randomly fail, leaving the websocket connection open, but the user disconnected.

My other concerns are latency and bandwidth. I have a goal for netplay to be optimized as much as possible, to be playable on 3G speeds, and to not have to have the worry of any data cap, which is why I haven't gone the approach of streaming the video.

Our current take on netplay is this. We're planning on basically reverse engineering how netplay works in retroarch and putting it in here. I think the main issue was that game state prediction is unimplimented at the moment, after that we should be able to optimize latency and delay as much as possible, the same way retroarch does it. @allancoding is assigned with this task, and I got no clue where the progress is at.


Now my question, how did you get an audio stream again? I've tried and tried but I've never been able to get an audio stream. Could you try recording it with a mediarecorder? Because that's my current goal that I haven't been able to accomplish

**Comment by n-at** [comment]
About audio capture. It's code here: https://github.com/n-at/playtime/blob/master/assets/netplay.js#L394 AL exposes `GainNode` in `AL.currentCtx.audioCtx.sources` - in EmulatorJS it used to control game volume, but it's output can also be connected to something else. I've connected it to the `ChannelMergerNode` (in case if there is multiple gain nodes) and merger's output connected to the `MediaStreamAudioDestinationNode ` - it produces `MediaStream` that can be sent via WebRTC to clients. I think that the merger node is not really needed, and there will always be only one GainNode (so code may be simplified). There is one rare glitch - sometimes audio stream is empty and I have not yet found a cause (maybe audio capture started too early and not initialized properly).

`MediaRecorder` works. I just took function from netplay (with little edit) and passed collected MediaStream to recorder:

```js
function collectMediaTracks() {
        let videoTrack = null;
        const videoTracks = document.querySelector(`${window.EJS_player} canvas`).captureStream().getVideoTracks();
        if (videoTracks.length !== 0) {
            videoTrack = videoTracks[0];
        } else {
            console.error('Unable to capture video stream');
            return null;
        }

        let audioTrack = null;
        if (window.AL && window.AL.currentCtx && window.AL.currentCtx.audioCtx) {
            const alContext = window.AL.currentCtx;
            const audioContext = alContext.audioCtx;

            const gainNodes = [];
            for (let sourceIdx in alContext.sources) {
                gainNodes.push(alContext.sources[sourceIdx].gain);
            }

            const merger = audioContext.createChannelMerger(gainNodes.length);
            gainNodes.forEach(node => node.connect(merger));

            const destination = audioContext.createMediaStreamDestination();
            merger.connect(destination);

            const audioTracks = destination.stream.getAudioTracks();
            if (audioTracks.length !== 0) {
                audioTrack = audioTracks[0];
            }
        }

        const stream = new MediaStream();
        if (videoTrack && videoTrack.readyState === 'live') {
            stream.addTrack(videoTrack);
        }
        if (audioTrack && audioTrack.readyState === 'live') {
            stream.addTrack(audioTrack);
        }
        return stream;
    }

    setTimeout(() => {
        const chunks = [];
        const recorder = new MediaRecorder(collectMediaTracks());
        recorder.addEventListener('dataavailable', e => {
            chunks.push(e.data);
        });
        recorder.addEventListener('stop', () => {
            const blob = new Blob(chunks);
            const url = URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = 'video.webm';
            a.click();
        });
        recorder.start();

        setTimeout(() => { recorder.stop()}, 10000);
    }, 5000);
```

And this video been recorded:

[video.webm](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/2debe9b5-4062-42a1-844c-db27789a5344)

Video looks fine but the game canvas refresh rate reduced to 1-2 FPS during recording. Maybe the canvas should not be captured directly (for example use another invisible canvas of fixed size and paint game canvas image on it).

**Reply by n-at** [reply]
About connection issues. Yes, sometimes RTC and WebSocket connection fails. I've added auto reconnection and it works. Also I've added heartbeats - from client to server via WebSocket and from client to game host via RTC Data Channel. If client is not responding to heartbeat in 10 seconds (interval picked randomly) it will be disconnected by server. Heartbeat messages now is empty, but RTC connection info can be added (RTC has some quality metrics) and if client says that game host or other clients not responding or ping is too high, server may disconnect it.

The main problem with WebRTC is connection speed. By my estimation, for one client 1 Mb/s will be needed for good video quality. If connection speed drops, video quality also drops (it is may be good for video calls but not for the gameplay). I've tested on LTE and it works good.

**Reply by n-at** [reply]
Game host performance is need to be tested, can't say much about it for now. On devices I've tested (not many) it works good.

**Reply by allancoding** [reply]
It seems to me like this might have better performance. Because right now what I have been working on is very complicated and requires both rewind and fast-forward. I think I could implement this better for more than one client by off loading so of the work to the server. By having the video and audio set over websocket to the server for processing and the server just making a video stream that the clients can just (watch). Sorry to repeat the question, but do you have discord or some other way of communication. I would love to have your help with this.
Thanks

**Reply by ethanaobrien** [reply]
@n-at how much bandwidth does this use? Do know know if there's a way to "lower streaming quality"?

**Reply by allancoding** [reply]
What do you mean bandwidth? webRTC is peer to peer. Yes the server could compress the video and lower the quality.

**Reply by ethanaobrien** [reply]
Peer to peer still goes over the internet

**Reply by allancoding** [reply]
The host would need to have a better device. But you could also compress the video and audio be for it is sent depending on the hosts connection.

**Reply by allancoding** [reply]
> Peer to peer still goes over the internet

Sorry I thought you meant server bandwidth. Probably less than if you were watching a video on YouTube or something similar.

**Reply by n-at** [reply]
I don't have a discord account (only email from github profile).

Bandswidth (my estimation) for the game host - 1 Mb/s per client, for the client - 1 Mb/s total (but it can adapt to current connection quality). RTC video and audio streams quality can be controlled via https://developer.mozilla.org/en-US/docs/Web/API/RTCRtpSender/setParameters (set max bitrate, max framerate, scale resolution).

**Reply by allancoding** [reply]
The problem is the more clients there are the more, the host computer has to work. I think streaming the connection from a server would be easier on the host.

**Reply by ethanaobrien** [reply]
Hmm, the ability to change the bitrate seems perfect.

>The problem is the more clients there are the more, the host computer has to work. I think streaming the connection from a server would be easier on the host.

No no no, we aren't doing any streaming from a server. This isn't a cloud streaming service. It shouldn't have to work too hard


@n-at would you like to try integrating this with EmulatorJS netplay? Stripping out (what you need)? If you have any questions or need access to the current emulatorjs netplay repo (which I expect) just let me know.

Also, do we need a webrtc turn server for this?

**Reply by allancoding** [reply]
I would not "stream" it would just manage sending the video to the clients instead of the host sending the video individually.

**Reply by ethanaobrien** [reply]
I still think that would put too much work on the server. It shouldn't make a noticeable difference on the host

**Reply by allancoding** [reply]
I would like to try it without webRTC. We should try testing it and the performance hit. I would think that the server would not have to do too much because it would just be routing data no processing.

**Reply by ethanaobrien** [reply]
I'd like @n-at to try integrating this into emulatorjs and see how it performs. At the very least it will be netplay for a while

As you said @allancoding , webrtc is peer to peer. The data doesn't go through the server

**Reply by allancoding** [reply]
depending on the host it will have to make a peer to peer connection to each client individually.

**Reply by ethanaobrien** [reply]
What

**Reply by n-at** [reply]
> @n-at would you like to try integrating this with EmulatorJS netplay? Stripping out (what you need)? If you have any questions or need access to the current emulatorjs netplay repo (which I expect) just let me know.

Sorry, but I couldn't do this due to my current workload. Appreciate your understanding.

> Also, do we need a webrtc turn server for this?

Yes, WebRTC requires a TURN/STUN/ICE server to connect peers over the internet.

**Reply by ethanaobrien** [reply]
>Sorry, but I couldn't do this due to my current workload. Appreciate your understanding.

Thats alright, no problem there. I did get some other idea's I'd like to play around with for netplay so I'll let you know how those go

**Reply by n-at** [reply]
> Video looks fine but the game canvas refresh rate reduced to 1-2 FPS during recording. Maybe the canvas should not be captured directly (for example use another invisible canvas of fixed size and paint game canvas image on it).

Capture video with additional canvas works. Game frame rate is good and video is recording.

```js
function collectMediaTracks(canvasEl, fps) {
        let videoTrack = null;
        const videoTracks = canvasEl.captureStream(fps).getVideoTracks();
        if (videoTracks.length !== 0) {
            videoTrack = videoTracks[0];
        } else {
            console.error('Unable to capture video stream');
            return null;
        }

        let audioTrack = null;
        if (window.AL && window.AL.currentCtx && window.AL.currentCtx.audioCtx) {
            const alContext = window.AL.currentCtx;
            const audioContext = alContext.audioCtx;

            const gainNodes = [];
            for (let sourceIdx in alContext.sources) {
                gainNodes.push(alContext.sources[sourceIdx].gain);
            }

            const merger = audioContext.createChannelMerger(gainNodes.length);
            gainNodes.forEach(node => node.connect(merger));

            const destination = audioContext.createMediaStreamDestination();
            merger.connect(destination);

            const audioTracks = destination.stream.getAudioTracks();
            if (audioTracks.length !== 0) {
                audioTrack = audioTracks[0];
            }
        }

        const stream = new MediaStream();
        if (videoTrack && videoTrack.readyState === 'live') {
            stream.addTrack(videoTrack);
        }
        if (audioTrack && audioTrack.readyState === 'live') {
            stream.addTrack(audioTrack);
        }
        return stream;
    }

    const CAPTURE_SCREEN_WIDTH = 800;
    const CAPTURE_SCREEN_HEIGHT = 600;
    const CAPTURE_FPS = 30;

    const captureCanvas = document.createElement('canvas');
    captureCanvas.width = CAPTURE_SCREEN_WIDTH;
    captureCanvas.height = CAPTURE_SCREEN_HEIGHT;
    captureCanvas.style.position = 'absolute';
    captureCanvas.style.top = '-999px';
    captureCanvas.style.bottom = '-999px';
    document.getElementsByTagName('body')[0].append(captureCanvas);

    const captureCtx = captureCanvas.getContext('2d');
    captureCtx.fillStyle = '#000';

    setTimeout(() => {
        const gameCanvas = document.querySelector(`${window.EJS_player} canvas`);
        const interval = setInterval(() => {
            const scaleX = CAPTURE_SCREEN_WIDTH / gameCanvas.width;
            const scaleY = CAPTURE_SCREEN_HEIGHT / gameCanvas.height;
            const scale = Math.max(scaleY, scaleX);
            const width = gameCanvas.width * scale;
            const height = gameCanvas.height * scale;
            const startX = (CAPTURE_SCREEN_WIDTH - width) / 2;
            const startY = (CAPTURE_SCREEN_HEIGHT - height) / 2;
            captureCtx.drawImage(gameCanvas, startX, startY, width, height);
        }, 1000 / 60);

        const chunks = [];
        const recorder = new MediaRecorder(collectMediaTracks(captureCanvas, CAPTURE_FPS), {
            videoBitsPerSecond: 2 * 1024 * 1014,
            audioBitsPerSecond: 256 * 1024,
        });
        recorder.addEventListener('dataavailable', e => {
            chunks.push(e.data);
        });
        recorder.addEventListener('stop', () => {
            const blob = new Blob(chunks);
            const url = URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = 'video.webm';
            a.click();

            clearInterval(interval);
            captureCanvas.remove();
        });
        recorder.start();

        setTimeout(() => { recorder.stop()}, 10000);
    }, 5000);
```

[video (9).webm](https://github.com/EmulatorJS/EmulatorJS/assets/2185280/c2428c01-30b6-4104-908f-57ee55501910)

**Comment by Cattn** [comment]
I also like this method of netplay, and some friend of mine & me have made a similar solution in the past (albiet much much more crude lol as it was meant for only about 4 of us), but without webRTC (due to complications)

Was very little compared to this
